aptos-move/mvhashmap/src/versioned_group_data.rs

```diff
@@ -1,473 +1,249 @@
-        // Remove idx updates first, then entries.
-        let idx_update_tags: Vec<(T, Option<usize>)> = self
-            .idx_to_update
-            .remove(&shifted_idx)
-            .map_or(vec![], |map| {
-                map.into_inner()
-                    .into_iter()
-                    .map(|(tag, v)| (tag, v.bytes_len()))
-                    .collect()
-            });
-
-        // Similar to mark_estimate, need to remove an individual entry for each tag.
-        for (tag, _) in idx_update_tags.iter() {
-            assert_some!(
-                self.versioned_map
-                    .get_mut(tag)
-                    .expect("Versioned entry must exist for tag")
-                    .remove(&shifted_idx),
-                "Entry for tag / idx must exist to be removed"
-            );
-        }
-
-        idx_update_tags
-    }
-
-    // Records the latest committed op for each tag in the group (removed tags ar excluded).
-    fn commit_idx(
-        &mut self,
-        shifted_idx: ShiftedTxnIndex,
-        allow_new_modification: bool,
-    ) -> anyhow::Result<()> {
-        use std::collections::hash_map::Entry::*;
-        use WriteOpKind::*;
-
-        let idx_updates = self
-            .idx_to_update
-            .get(&shifted_idx)
-            .expect("Group updates must exist at the index to commit");
-        for (tag, v) in idx_updates.iter() {
-            match (self.committed_group.entry(tag.clone()), v.write_op_kind()) {
-                (Occupied(entry), Deletion) => {
-                    entry.remove();
-                },
-                (Occupied(mut entry), Modification) => {
-                    entry.insert(v.clone());
-                },
-                (Vacant(entry), Creation) => {
-                    entry.insert(v.clone());
-                },
-                (Vacant(entry), Modification) if allow_new_modification => {
-                    entry.insert(v.clone());
-                },
-                (Occupied(mut entry), Creation) if entry.get().write_op_kind() == Deletion => {
-                    entry.insert(v.clone());
-                },
-                (e, _) => {
-                    bail!(
-                        "[{shifted_idx:?}] WriteOp kind {:?} not consistent with previous value at tag {tag:?}, value: {e:?}",
-                        v.write_op_kind(),
-                    );
-                },
             }
         }
 
         Ok(())
     }
 
-    fn get_committed_group(&self) -> Vec<(T, ValueWithLayout<V>)> {
-        self.committed_group.clone().into_iter().collect()
+    pub fn update_tagged_base_value_with_layout(
+        &self,
+        group_key: K,
+        tag: T,
+        value: V,
+        layout: Option<Arc<MoveTypeLayout>>,
+    ) {
+        self.values.set_base_value(
+            (group_key, tag),
+            ValueWithLayout::Exchanged(Arc::new(value), layout.clone()),
+        );
     }
 
-    fn get_latest_tagged_value(
+    pub fn write(
```

Comment:
Can we have these functions documented so that there is no need to check the impl to understand why bool is returned, etc?

Reply:
added some doc, also to address @zekun000's concern naming "changes_behavior". It's just ret now lol